cpp/include/cudf/io/text/detail/multistate.hpp

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cstdint>
+
+namespace cudf {
+namespace io {
+namespace text {
+namespace detail {
+
+/**
+ * @brief represents a single (begin, end] pair of possible state transition history.
+ *
+ */
+struct multistate_segment {
+ public:
+  static auto constexpr max_states = 16;
+  constexpr multistate_segment() : _data(0) {}
+  constexpr multistate_segment(uint8_t head, uint8_t tail) : _data((head & 0b1111) | (tail << 4)) {}
+
+  constexpr uint8_t get_head() const { return _data & 0b1111; }
+  constexpr uint8_t get_tail() const { return _data >> 4; }
+
+ private:
+  uint8_t _data;
+};
+
+/**
+ * @brief Holds up to 7 transition history segments
+ */
+struct multistate {
+ public:
+  static auto constexpr max_segments = 7;
```

Comment:
I guess this is just for performance reasons, such that you can still fit a `multistate` within eight bytes? Just for my understanding, is there anything particularly bad happening if we'd have more than `7` `max_segments`?

Reply:
have not tested for more than 7 segments, so can't say, but the idea was to fit within 8 bytes.